### PR TITLE
feat(app-router): introduce `experimental.missingSuspenseWithCSRBailout` flag

### PR DESCRIPTION
**Pull request by feedthejim**
### What?

This PR adds a new flag called `experimental.missingSuspenseWithCSRBailout`.

### Why?

Via this PR we can break a build when calling `useSearchParams` without wrapping it in a suspense boundary.

If no suspense boundaries are present, Next.js must avoid doing SSR and defer the entire page's rendering to the client. This is not a great default. Instead, we will now break the build so that you are forced to add a boundary.

### How?

Add an experimental flag. If a `BailoutToCSRError` error is thrown and this flag is enabled, the build should fail and log an error, instead of showing a warning and bail the entire page to client-side rendering.

Closes NEXT-1770

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 3477c765561e6cb10bd77218d8ad346ca6e80079

`pnpm test-dev test/development/basic/gssp-ssr-change-reloading/test/index.test.ts` 
- GS(S)P Server-Side Change Reloading > should not reload page when client-side is changed too GSSP

<details>
<summary>Expand output</summary>

  ● GS(S)P Server-Side Change Reloading › should not reload page when client-side is changed too GSSP

    expect(received).toBe(expected) // Object.is equality

    Expected: "hi"
    Received: undefined

      217 |
      218 |     await check(() => browser.elementByCss('#change').text(), 'changed')
    > 219 |     expect(await browser.eval(`window.beforeChange`)).toBe('hi')
          |                                                       ^
      220 |
      221 |     const props2 = JSON.parse(await browser.elementByCss('#props').text())
      222 |     expect(props).toEqual(props2)

      at Object.toBe (development/basic/gssp-ssr-change-reloading/test/index.test.ts:219:55)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | vercel/next.js feedthejim/forbid-no-suspense | Change |
| - | - | - | - |
| buildDuration | 14.9s | 14.6s | N/A |
| buildDurationCached | 8.3s | 8.5s | ⚠️ +180ms |
| nodeModulesSize | 175 MB | 175 MB | N/A |
| nextStartRea..uration (ms) | 494ms | 494ms | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js feedthejim/forbid-no-suspense | Change |
| - | - | - | - |
| 199-HASH.js gzip | 32.3 kB | 32.4 kB | N/A |
| 3f784ff6-HASH.js gzip | 53.2 kB | 53.2 kB | ✓ |
| 99.HASH.js gzip | 182 B | 182 B | ✓ |
| framework-HASH.js gzip | 45.5 kB | 45.5 kB | ✓ |
| main-app-HASH.js gzip | 254 B | 252 B | N/A |
| main-HASH.js gzip | 35.3 kB | 35.4 kB | N/A |
| webpack-HASH.js gzip | 1.75 kB | 1.75 kB | N/A |
| Overall change | 98.9 kB | 98.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js feedthejim/forbid-no-suspense | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js feedthejim/forbid-no-suspense | Change |
| - | - | - | - |
| _app-HASH.js gzip | 206 B | 205 B | N/A |
| _error-HASH.js gzip | 182 B | 180 B | N/A |
| amp-HASH.js gzip | 506 B | 505 B | N/A |
| css-HASH.js gzip | 322 B | 323 B | N/A |
| dynamic-HASH.js gzip | 2.59 kB | 2.59 kB | ✓ |
| edge-ssr-HASH.js gzip | 260 B | 259 B | N/A |
| head-HASH.js gzip | 350 B | 350 B | ✓ |
| hooks-HASH.js gzip | 369 B | 369 B | ✓ |
| image-HASH.js gzip | 4.38 kB | 4.38 kB | N/A |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 2.67 kB | 2.67 kB | N/A |
| routerDirect..HASH.js gzip | 316 B | 318 B | N/A |
| script-HASH.js gzip | 385 B | 384 B | N/A |
| withRouter-HASH.js gzip | 319 B | 319 B | ✓ |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 3.99 kB | 3.99 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js feedthejim/forbid-no-suspense | Change |
| - | - | - | - |
| _buildManifest.js gzip | 484 B | 482 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js feedthejim/forbid-no-suspense | Change |
| - | - | - | - |
| index.html gzip | 529 B | 530 B | N/A |
| link.html gzip | 542 B | 542 B | ✓ |
| withRouter.html gzip | 524 B | 525 B | N/A |
| Overall change | 542 B | 542 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js feedthejim/forbid-no-suspense | Change |
| - | - | - | - |
| edge-ssr.js gzip | 96.1 kB | 96.1 kB | N/A |
| page.js gzip | 140 kB | 140 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js feedthejim/forbid-no-suspense | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 624 B | 624 B | ✓ |
| middleware-r..fest.js gzip | 150 B | 151 B | N/A |
| middleware.js gzip | 25.8 kB | 25.8 kB | N/A |
| edge-runtime..pack.js gzip | 1.92 kB | 1.92 kB | ✓ |
| Overall change | 2.55 kB | 2.55 kB | ✓ |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>page.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 3477c765561e6cb10bd77218d8ad346ca6e80079

File: packages/next/src/server/app-render/app-render.tsx [hunk 1]
```diff
@@ -788,10 +788,11 @@ async function renderToHTMLOrFlightImpl(
         }
 
         if (err.digest === NEXT_DYNAMIC_NO_SSR_CODE) {
-          warn(
+          err(
             `Entire page ${pagePath} deopted into client-side rendering. https://nextjs.org/docs/messages/deopted-into-client-rendering`,
             pagePath
           )
```
Comment (SukkaW) [inline packages/next/src/server/app-render/app-render.tsx:794]:
```suggestion
          error(
            `Entire page ${pagePath} deopted into client-side rendering. https://nextjs.org/docs/messages/deopted-into-client-rendering`,
            pagePath
          )
```

File: packages/next/src/shared/lib/lazy-dynamic/dynamic-bailout-to-csr.tsx [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+import type { ReactElement } from 'react'
+import { BailoutToCSRError } from './bailout-to-csr'
+
+interface BailoutToCSRProps {
+  reason: string
+  children: ReactElement
+}
+
+export function BailoutToCSR({ reason, children }: BailoutToCSRProps) {
```
Comment (wyattjoh) [inline packages/next/src/shared/lib/lazy-dynamic/dynamic-bailout-to-csr.tsx:11]: Add some comments explaining the acronym.

**Review by wyattjoh — APPROVED** [review]

**Pull request merged**